### ISSUE TITLE
blockchain.info removed the "Unspent outputs" link

### ISSUE THREAD

**Issue by bitcoinhodler** [GlacierProtocol/GlacierProtocol#49]
The first steps of the Glacier withdrawal protocol involve finding the UTXOs to spend, by using blockchain.info. Unfortunately this service has apparently removed the "Unspent outputs" link that Glacier instructs us to use.

According to their [API docs](https://www.blockchain.com/api/blockchain_api), you can get the list of UTXOs for an address in an ugly format using their API: `https://blockchain.info/unspent?active=$address`

Fortunately the raw transaction API (`https://blockchain.info/tx/transaction-id-here?format=hex`) still seems to work.

**Comment by bitcoinhodler** [comment]
Note that with the `https://blockchain.info/unspent?active=$address` API, the txid is the key labeled `"tx_hash_big_endian"`.

**Comment by bitcoinhodler** [comment]
I've [requested](https://github.com/Blockstream/esplora/issues/33) blockstream.info to add this to their service, and then we can switch from blockchain.info to blockstream.info with only a few tweaks to the Glacier document.

**Comment by shesek** [comment]
I made a little in reply to @bitcoinhodler's [feature request](https://github.com/Blockstream/esplora/issues/33) that you might find this useful: https://codepen.io/anon/pen/YBRder.

You should ideally open this as a local HTML page rather than from codepen.

**Comment by bitcoinhodler** [comment]
@shesek this looks handy. Is it something we could easily host here on glacierprotocol.org? If we commit the generated HTML into Git?

Feature request: for each txid, add a link to generate a QR code from the raw transaction hex. (This could be complicated because a large tx will require multiple QR codes, to be recombined by the user.)

**Comment by shesek** [comment]
> Is it something we could easily host here on glacierprotocol.org? If we commit the generated HTML into Git?

Yes, you definitely could. I would also remove the dependency on jQuery to make this more self-contained. A single `.html` file with everything embedded in it is the most ideal.

Also, maybe advice users to run this html page locally? That way the potential privacy leak is limited just to blockstream's servers and not also to glacier's (the page shouldn't be sending anything to glacier's servers, but the server could modify the javascript code to send data to it anyway).

Another good thing to suggest is using Tor and the `.onion` hidden service. You could have a separate `.html` page that queries [blockstream.info's hidden service v3](http://explorerzydxu5ecjrkwceayqybizmpjjznk5izmitf2modhcusuqlid.onion/) instead of `https://blockstream.info/`.

But really, users should be getting this information from their own full node...

> Feature request: for each txid, add a link to generate a QR code from the raw transaction hex.

I'll look into this. This does it make it harder (impossible really) to have this as a standalone script with no dependency on libraries.

A starting point could be to add a link to the rawtx in hex, which is available at `https://blockstream.info/api/tx/<txid>/hex` ([example](https://blockstream.info/api/tx/300f4d2ba943c2e1bf8dd815486332f5983e6d3d3c2460e37072780d3e5d9a4e/hex)).

Users could then generate QRs using something like `qtqr` (a GUI for generating QRs, available in debian's/ubuntu's repos) or `qrencode` (a CLI tool for generating QRs, also available in the repos).

**Comment by shesek** [comment]
Btw, why do users need to get the full previous transaction of the outputs they're spending? Isn't that no longer really necessary, now that we have segwit and signing over input amounts?

**Comment by shesek** [comment]
I noticed that `glacierprotocol.org` embeds Google Analytics's tracking code and Google's Web Fonts, please make sure the UTXO extraction tool page doesn't do that :)

**Comment by bitcoinhodler** [comment]
> 
> 
> Btw, why do users need to get the full previous transaction of the outputs they're spending? Isn't that no longer really necessary, now that we have segwit and signing over input amounts?

Glacier predates Segwit, and only the latest cold-storage addresses created with Glacier will be segwit-enabled. It's far simpler to use the same flow for every address, versus two different flows for the two possible types of addresses.

I'd be grateful if you would file an issue on [glacierprotocol.github.io](https://github.com/GlacierProtocol/glacierprotocol.github.io) (or even better, a PR) to remove the Google tracking crap. I'm not sure if that was intentional or accidental but I personally don't like it there.

**Comment by bitcoinhodler** [comment]
> Also, maybe advice users to run this html page locally? That way the potential privacy leak is limited just to blockstream's servers and not also to glacier's (the page shouldn't be sending anything to glacier's servers, but the server could modify the javascript code to send data to it anyway).

How about we include the .html file in the download package, and instruct users to launch it from their local copy?

> But really, users should be getting this information from their own full node...

Yep, see #54.

> A starting point could be to add a link to the rawtx in hex, which is available at `https://blockstream.info/api/tx/<txid>/hex` ([example](https://blockstream.info/api/tx/300f4d2ba943c2e1bf8dd815486332f5983e6d3d3c2460e37072780d3e5d9a4e/hex)).

Yes, that would be great. The protocol currently has users manually construct such URLs and then paste the raw hex into a QR generator, so this would simplify that flow a bit.

**Comment by bitcoinhodler** [comment]
> Another good thing to suggest is using Tor and the `.onion` hidden service. You could have a separate `.html` page that queries [blockstream.info's hidden service v3](http://explorerzydxu5ecjrkwceayqybizmpjjznk5izmitf2modhcusuqlid.onion/) instead of `https://blockstream.info/`.

I've had trouble in the past trying to load files from the local filesystem in the Tor Browser. It really doesn't want to do that. The proposed solution? Use Firefox for that. But then, Firefox can't access the onion API. So I have a suspicion that supporting Tor in this way would be difficult.

**Comment by shesek** [comment]
> I'd be grateful if you would file an issue on [glacierprotocol.github.io](https://github.com/GlacierProtocol/glacierprotocol.github.io)

--> https://github.com/GlacierProtocol/glacierprotocol.github.io/issues/31

> How about we include the .html file in the download package, and instruct users to launch it from their local copy?

Sounds good.

> I've had trouble in the past trying to load files from the local filesystem in the Tor Browser.

You can use Python's built-in http server module to start a static web server by running `python3 -m http.server` (for python 3) or `python -m SimpleHTTPServer` (for python 2) inside the directory with the files.

But then again, I'm not really sure if the Tor browser bundle likes accessing `localhost`? I never tried that.

**Comment by bitcoinhodler** [comment]
> Edit: yes, it doesn't like that, obviously... I'll try to see if I can figure something else out.

I wouldn't worry about Tor for the time being. Privacy is not a primary goal of Glacier. Getting the UTXOs is completely broken today, and is therefore urgent. Don't hold it up for Tor.

**Comment by shesek** [comment]
Here's a jQuery-free version as a single plain `.html` file: https://gist.github.com/shesek/f0f73fb74b0083c46cf625570536e7aa

It looks kinda bad without Bootstrap's CSS (which the previous codepen had), you can either consider adding all of it (I guess adding third-party CSS isn't too terrible, but it should be embedded into the html file itself and not referenced externally), or adding some custom simpler CSS instead. I would go with the latter option.

**Comment by bitcoinhodler** [comment]
This is very nice...though I agree it needs a little CSS love. Also I think the CSV table is not useful for Glacier users.

Can you add a footer identifying where this is getting its info? "Data sourced from blockstream.info"

Also I worry about the security impact of this. Currently Glacier does not ask the user to run any Glacier-provided code on any online PC. With this change, we would be running Glacier-provided JS code on their PC. For that reason alone, I would prefer to send the user to blockstream.info directly.
